proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,281 @@
+
+The application can be made smaller by support aggressive linking behavior (e.g. "tree-shaking"). This can be done in a number of ways outside the scope of this document. However, approaches in this document that do code-gen may automatically handle this due to generated calls to members.
+
+# Serializer design background
+Today the serializer avoids using Reflection.Emit on non-.NET scenarios through the use of an #ifdef. However, the serializer still uses emit indirectly through `Type.MakeGenericType`. Note that Xamarin stubs `MakeGenericType` and tries to support these on iOS by pre-generating types via AOT but there are still issues with this (including not being able to determine all possible `<T>` variants).
+
+The existing serializer makes use of custom converters which is a class that knows how to implement a Read(...) and Write(...) method. There are 3 types of converters:
+1. Value. The only type of custom converters that can be implemented publically today (by the community).
+2. Collection. Used internally for IEnumerable Types.
+3. Object. Used internally for all other Types.
+
+The converters are stored on the `JsonSerializerOptions` in a dictionary keyed on the Type to convert. A single converter may know how to convert multiple Types.
+
+Internal converters use metadata for a given Type which is obtained from using reflection (to look for properties and JsonAttribute-derived attributes on a Type and its members). This metadata is then merged with values from the `JsonSerializerOptions` instance which is possible since options instance is immutable once the first (de)serialization occurs.
+
+The immutable semantics of the options also help in other ways:
+- The metadata could be embedded into code generation (to avoid the reflection cost of looking up the metadata).
+- The metadata could be used to control how code is generated if the generated code has branching or other logic that can be determined based upon the metadata (for example, having a custom property lookup mechanism for each Type).
+- Avoids locks and issues that would occur if another thread changes the options.
+
+A custom Value converter (implemented by the community) does not have access to the metadata since the metadata types are internal. This has been an issue for some scenarios and opening up the metadata APIs has been requested.
+
+The built-in converters do have access to metadata. The metadata is maintained internally by a `JsonClassInfo` object for every type. For Object converters (meaning non-Value and non-Collection converters) there is also an instance of a `JsonPropertyInfo` object for every property.
+
+# Startup performance costs
+The overhead of deserializing a simple 4-property POCO is around **22ms** for the first run (<1ms for second run). This was measured in 5.0 master as of 3/27/2020 and tested on an Intel i7 3.2GHz.
```

Comment:
Link to GH hast instead of `was measured in 5.0 master as of 3/27/2020 `

Reply:
I'll link that the next time I run the benchmarks.